### PR TITLE
Replace qalloc with boost::pmr

### PR DESCRIPTION
**Pull request by seelabs**
## High Level Overview of Change

Replace qalloc with boost's `monotonic_memory_resource`. `std::pmr` could not be used as clang's standard library does not yet support it.

Note: `qalloc` is battle-tested and works well. One thing reviewers should consider is whether it's worth it to replace that code.

### Context of Change

This is a refactoring that replaces code with the equivalent functionality from boost.

### Type of Change

- [x ] Refactor (non-breaking change that only restructures code)

## Future Tasks

Replace this with `std::pmr` once clang's standard library supports `pmr` (microsoft and gcc already support it).

### CODE REVIEW

**Comment by scottschurr** [comment]
FWIW, I think this is a change worth making.  Just because something is battle tested does not make it maintenance free.  So I think it's a net win to replace our home-grown `qalloc` with a facility that we think will see long term support in the standard library.  I think it's okay using boost as an interim solution before transitioning to using the standard library.

However I am not a fan of replacing `std::pair` with `std::tuple`.  We should replace the `pair`s with aggregates.  In my opinion the use of `pair` in those situations was questionable anyway.  Here's what it looks like if we use aggregates instead: https://github.com/scottschurr/rippled/commit/aaf16763608061fc19965b29d5d651a4440a1ce7  I think it's an improvement.

**Comment by seelabs** [comment]
I don't love `pairs` and `tuples` myself and also prefer to replace them with structs - nice change @scottschurr, cherry picked (although I do consider `pairs` and `tuples` equivalent; one is not worse than the other).

**Comment by scottschurr** [comment]
Thanks for the cherry-pick.

Just to put forward my own opinion, I consider `pair` and `tuple` to be conceptually equivalent, but not functionally equivalent.  Reading `first` and `second` to access the elements of a pair is (to me) slightly more intuitive than reading `std::get<0>(tupleName)>`.  So I think that `tuple` is slightly worse than `pair` for readability.  But having actual useful names for the members (with an aggregate) usually beats them both.

**Comment by cjcobb23** [comment]
> Thanks for the cherry-pick.
> 
> Just to put forward my own opinion, I consider `pair` and `tuple` to be conceptually equivalent, but not functionally equivalent. Reading `first` and `second` to access the elements of a pair is (to me) slightly more intuitive than reading `std::get<0>(tupleName)>`. So I think that `tuple` is slightly worse than `pair` for readability. But having actual useful names for the members (with an aggregate) usually beats them both.

I agree with this sentiment, that `tupleName.first` is a lot more readable than `std::get<0>(tupleName)`

**Comment by scottschurr** [comment]
I used running all of the unit tests single threaded as a way to do some very rough timings.  I used...

- develop (which uses `qalloc`) and
- this pull request. 

And, just to get a base line case, I also ran the tests after

- replacing the two uses of `qalloc` in develop with the `std::allocator`.

I ran the unit tests 5 times for each case and got the following results:
```
     timings of all unit tests single-threaded
std::allocator    qalloc    monotonic_buffer_resource
        789.4s    425.4s    435.0s
        783.6s    446.1s    449.7s
        779.1s    427.6s    451.9s
        775.6s    436.3s    455.2s
        768.3s    430.6s    439.5s

avg:    779.2s    433.2s    446.3s
          100%     55.6%     57.3%
```
The timings suggest that using `monotonic_buffer_resource` may be _slightly_ slower than using `qalloc`.  I don't think that possible slowdown is a good reason to throw out this pull request.  Sometimes using a library implementation is a little slower than a hand crafted solution and it's still worth living with that slight slowdown.

But I have an (unproven) hypothesis regarding the possible slight slowdown.  This pull request not only uses `boost::pmr::monotonic_buffer_resource`, it also uses `boost::pmr::polymorphic_allocator`.  The `polymorphic_allocator` uses type erasure, so it adds an extra level of indirection over old-style non-polymorphic allocators.  And I don't believe we're really making use of the polymorphic aspect of `polymorphic_allocator`.

I _wonder_ if replacing `polymorphic_allocator` with a custom (non-polymorphic) allocator that uses `monotonic_buffer_resource` as its backing store might get us that little bit of performance back?  My understanding is that the technical debt of a custom allocator should be a lot smaller than the technical debt of carrying around `qalloc`.  So I think, even with the custom allocator, we'd have a net win.  That's assuming that the custom allocator gets us back that (possible) small performance loss.

Thoughts?  I could easily be jumping to too many conclusions based on questionable information...

**Comment by seelabs** [comment]
@scottschurr Thanks for running that analysis, good stuff! I'm actually surprised that there's a slight slowdown over qalloc (tho it's small enough that I'm not that worried about it). Replacing the `polymorphic_allocator` is an interesting idea. One reason I'd like to keep it is I want to experiment with chaining a pooled allocator to a monotonic_buffer resource (on the theory it will use less memory, because freed memory would go back into the pool). I don't want to do that in this PR, but keeping the `polymorphic_allocator` keeps that option open.

**Review by scottschurr — APPROVED** [review]
:+1: I think removing the home-brew `qalloc` is a good idea.  I'm fine with you using `polymorphic_allocator`, since you have upcoming plans for the polymorphic aspect.  Thanks for doing this!

File: src/ripple/ledger/detail/RawStateTable.h [hunk 1]
```diff
@@ -35,8 +38,17 @@ class RawStateTable
...
-    RawStateTable() = default;
-    RawStateTable(RawStateTable const&) = default;
+    RawStateTable()
+        : monotonic_resource_{std::make_unique<
+              boost::container::pmr::monotonic_buffer_resource>(kilobytes(256))}
+        , items_{monotonic_resource_.get()} {};
+
+    RawStateTable(RawStateTable const& rhs)
+        : monotonic_resource_{std::make_unique<
+              boost::container::pmr::monotonic_buffer_resource>(kilobytes(256))}
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/ledger/detail/RawStateTable.h:48]: It's what qalloc defaulted to. I don't know where that number came from.

[the remark this answers is not in the document]
Comment (scottschurr) [reply src/ripple/ledger/detail/RawStateTable.h:48]: Fair.  I still think it's useful to leave a comment saying that no one knows how the initial value was determined due to murky history.  But I won't insist.

[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/ledger/detail/RawStateTable.h]: Fixed in `e5f630e158 [fold] Add initialBufferSize constant & misc changes:`

File: src/ripple/ledger/impl/OpenView.cpp [hunk 2]
```diff
@@ -89,7 +108,10 @@ OpenView::OpenView(
 }
 
 OpenView::OpenView(ReadView const* base, std::shared_ptr<void const> hold)
-    : rules_(base->rules())
+    : monotonic_resource_{std::make_unique<
+          boost::container::pmr::monotonic_buffer_resource>(kilobytes(256))}
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/ledger/impl/OpenView.cpp]: Fixed in `e5f630e158 [fold] Add initialBufferSize constant & misc changes:`

File: src/test/csf/Scheduler.h [hunk 3]
```diff
@@ -141,7 +144,7 @@ class Scheduler
         erase(iterator iter);
     };
 
-    qalloc alloc_;
+    boost::container::pmr::monotonic_buffer_resource alloc_{256 * 1024};
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/csf/Scheduler.h]: Fixed in `e5f630e158 [fold] Add initialBufferSize constant & misc changes:`

**Review by scottschurr — APPROVED** [review]
:+1: Looks great! Thanks.

**Review by HowardHinnant — APPROVED** [review]

**Pull request closed without merging**